airflow/operators/s3_to_redshift_operator.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from airflow.hooks.postgres_hook import PostgresHook
+from airflow.hooks.S3_hook import S3Hook
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+
+class S3ToRedshiftTransfer(BaseOperator):
+    """
+    Executes an COPY command to load files from s3 to Redshift
+
+    :param schema: reference to a specific schema in redshift database
+    :type schema: string
+    :param table: reference to a specific table in redshift database
+    :type table: string
+    :param s3_bucket: reference to a specific S3 bucket
+    :type s3_bucket: string
+    :param s3_key: reference to a specific S3 key
+    :type s3_key: string
+    :param redshift_conn_id: reference to a specific redshift database
+    :type redshift_conn_id: string
+    :param aws_conn_id: reference to a specific S3 connection
+    :type aws_conn_id: string
+    :param copy_options: reference to a list of COPY options
+    :type copy_options: list
+    """
+
+    template_fields = ()
+    template_ext = ()
+    ui_color = '#ededed'
+
+    @apply_defaults
+    def __init__(
+            self,
+            schema,
+            table,
+            s3_bucket,
+            s3_key,
+            redshift_conn_id='redshift_default',
+            aws_conn_id='aws_default',
+            copy_options=tuple(),
+            autocommit=False,
+            parameters=None,
+            *args, **kwargs):
+        super(S3ToRedshiftTransfer, self).__init__(*args, **kwargs)
+        self.schema = schema
+        self.table = table
+        self.s3_bucket = s3_bucket
+        self.s3_key = s3_key
+        self.redshift_conn_id = redshift_conn_id
+        self.aws_conn_id = aws_conn_id
+        self.copy_options = copy_options
+        self.autocommit = autocommit
+        self.parameters = parameters
+
+    def execute(self, context):
+        self.hook = PostgresHook(postgres_conn_id=self.redshift_conn_id)
+        self.s3 = S3Hook(aws_conn_id=self.aws_conn_id)
+        credentials = self.s3.get_credentials()
+        copy_options = '\n\t\t\t'.join(self.copy_options)
+
+        copy_query = """
+                        COPY {0}.{1}
+                        FROM 's3://{2}/{3}/{1}_'
```

Comment:
Can I ask what the purpose of the trailing underscore is?

Reply:
@artwr Thanks for the review and sorry for the late reply.
It corresponds to [the opposite operation](/apache/incubator-airflow/blob/b918a4976a641a3c944ef4ad4323fe25bd219ea0/airflow/operators/redshift_to_s3_operator.py#L96) (UNLOAD in RedshiftToS3Transfer), which outputs files named tablename + underscore + postfix that Redshift appends. For example, when RedshiftToS3Transfer unloads the table called foo, it will dump it to the files named foo_xxxx_part_xx. So `COPY schema.foo FROM 's3://bucket/prefix/foo_` loads these files to S3, since their paths are forwardly matched with the path specified in the FROM clause.

But I just noticed that in the case of using S3ToRedshiftTransfer solely, it's an unnecessary limitation that the input files must contain underscore in there name. And without that limitation, S3ToRedshiftTransfer works with the files dumped by RedshiftToS3Transfer. So I'll remove the underscore in the current PR.

Furthermore, these operators should be able to specify the table name and input file name independently. I'll file it as a new JIRA issue.

Reply:
👍